File: 2023/dart/lib/day8.dart
import 'package:aoc2023/advent_day.dart';
import 'package:aoc2023/util.dart';

main() => Day8().solve();

class Day8 extends AdventDay {
  Day8() : super(8);

  @override
  part1(String input) {
    final (instructions, nodes) = parseInput(input);

    var currentNode = "AAA";
    var iterations = 0;
    while (currentNode != "ZZZ") {
      final position = iterations % instructions.length;
      final lookup = nodes[currentNode];
      final instruction = instructions[position];
      currentNode = instruction == "L" ? lookup!.$1 : lookup!.$2;
      iterations++;
    }

    return iterations;
  }

  @override
  part2(String input) {
    final (instructions, nodes) = parseInput(input);

    var currentNodes = nodes.keys.where((key) => key.endsWith("A")).toList();
    var iterations = 0;
    var found = false;
    var r = List<int>.empty(growable: true);
    while (!found) {
      final position = iterations % instructions.length;
      iterations++;
      currentNodes = currentNodes
          .map((e) => nextNode(e, nodes, instructions[position]))
          .toList();
      final notEndingWithZ =
          currentNodes.where((e) => !e.endsWith("Z")).toList();
      if (notEndingWithZ.length != currentNodes.length) {
        r.add(iterations);
        currentNodes = notEndingWithZ;
      }
      if (notEndingWithZ.isEmpty) {
        found = true;
      }
    }

    return leastCommon(r);
  }

  leastCommon(List<int> numbers) {
    final start = numbers[0];
    final lcm = numbers.sublist(1).fold(start, (previousValue, element) {
      final z = gcd(previousValue, element);
      return (previousValue * element) ~/ z;
    });
    return lcm;
  }

  int gcd(int a, int b) {
    if (b == 0)
      return a;
    else
      return gcd(b, a % b);
  }

  getLeastCommon(n1, n2) {
    int max = n1 > n2 ? n1 : n2;
    while (true) {
      if (max % n1 == 0 && max % n2 == 0) {
        return max;
      }
      max++;
    }
  }

  String nextNode(String currentNode, Map<String, (String, String)> nodeMap,
      String instruction) {
    final lookup = nodeMap[currentNode];
    return instruction == "L" ? lookup!.$1 : lookup!.$2;
  }

  (String, Map<String, (String, String)>) parseInput(String input) {
    final parts = input.split("\n\n");
    final instructions = parts[0];
    final network = parts[1].rows();
    var nodes = Map<String, (String, String)>.fromIterable(network,
        key: (v) => v.split(" = ")[0],
        value: (v) {
          final lookups = v
              .split(" = ")[1]
              .replaceAll("(", "")
              .replaceAll(")", "")
              .split(", ");
          return (lookups[0], lookups[1]);
        });
    return (instructions, nodes);
  }
}

File: 2023/kotlin/src/main/kotlin/Day8.kt
fun main(args: Array<String>) {
    val input = readInput("input/day8")
    val day8 = Day8()
    day8.part2(input).println()
}

class Day8 {

    fun part2(input: String): Int {
        val p = input.split("\n\n");
        val instructions = p[0].trim().split("").drop(1).dropLast(1);
        val map = p[1].split("\n");

        val rows = HashMap<String, Lookup>();

        for (instruction in map) {
            val parts = instruction.split(" = ");
            println(parts[0] + " " + parts[1])
            val key = parts[0];
            val look = parts[1].replace("(", "").replace(")", "").split(", ");
            rows[key] = Lookup(look[0], look[1])
        }

        var iterations = 0;
        var found = false;
        var currentKey = map.first().split(" = ")[0]
        var turns = 0;

        while (!found) {
            val lookup = rows[currentKey]
            val instruction = instructions[iterations];
            val next = if(instruction == "L") lookup!!.left else lookup!!.right;
            if (next == "ZZZ") {
                found = true;
            }
            currentKey = next;
            iterations++;
            if (iterations == instructions.size) {
                turns++
                iterations = 0;
            }
        }

        return turns * instructions.size + iterations;
    }


}

data class Lookup(val left: String, val right: String)
